### ISSUE TITLE
Handling of `annot` argument in graph3D

### ISSUE THREAD

**Issue by thibautjombart** [reconhub/epicontacts#66]
From these lines:
https://github.com/reconhub/epicontacts/blob/master/R/graph3D.R#L103-L112

it looks like the function will only behave well as expected if a logical is given for `annot`, which does not agree with the doc (an index of columns in $linelist).

See behaviour in `vis_epicontacts`:
https://github.com/reconhub/epicontacts/blob/master/R/vis_epicontacts.R#L86-L93
and then
https://github.com/reconhub/epicontacts/blob/master/R/vis_epicontacts.R#L113-L118

@nistara do you think we can have something similar with `graph3D`? Incidentally, I think once this is done and unit-tested, we'll be at 100% coverage =D

**Comment by vpnagraj** [comment]
@thibautjombart so just to be clear ... the "annot" argument in `vis_epicontacts` should *not  be a logical ... it's a character vector of column names to annotate the nodes, right? or could the function accept either a logical or a character vector?

it looks like the param is `TRUE` in the default and example:

https://github.com/reconhub/epicontacts/blob/master/R/vis_epicontacts.R#L64-L72

**Comment by thibautjombart** [comment]
Thanks, doc was unclear. It should be any acceptable subsetting of the columns names of $linelist really.
 I clarified the doc at:
bf6a4f3e0ab67af15c1c942041a0a1a2f69283c7

**Comment by thibautjombart** [comment]
Poke @nistara ;) Let me know if I can help

**Comment by finlaycampbell** [comment]
Just realised I 'fixed' the documentation on graph3D without checking the issues first, stupid me. I'll try actually fixing the issue now.

**Comment by nistara** [comment]
@thibautjombart , sorry!!! I was working on it and didn't comment!! The thing is using all the columns for annotations when `annot = TRUE` made it difficult to see the graph at the same time, since they don't show up as nice as visNetworks. I'll put the code on a branch for everyone to see today.

**Comment by thibautjombart** [comment]
No probs ;) I suspected as much. Bha.. whatever works, we're close to release now anyway :)

**Comment by nistara** [comment]
I've just pushed a branch (`graph-annot`) to see how it might change things. (Here)[https://github.com/reconhub/epicontacts/commit/075a964f01d7dc160d4ff0569e5dc329bb9b9e2f#diff-cfdcb2a9b46e8d24b27a796e8629db26] are the differences. I basically modified @thibautjombart 's code and incorporated it in the function. One thing that struck me was that when igraph converts the `epicontacts` object into an igraph object, it adds an extra column, `name`, which I copy and put back as `id`. I'm going to see what happens if the linelist contains a `name` column to begin with!

**Comment by vpnagraj** [comment]
i believe i have the "annot" argument behaving consistently between `graph3D()` and `vis_epicontacts()` and coverage is at 100%

@nistara i'm going to close this issue but i see you're working on another branch now (`fix-as.igraph`) ... feel free to open another issue as needed

**Comment by nistara** [comment]
@vpnagraj, there's consistent behaviour between graph3D() and vis_epicontacts(), but actually printing all the column information is messing with the display, and also making the graph sluggish in changing info from one node to another. I'm making the default display the "group" ("id" in case group is not specified) unless the user specifies more columns, and adding that in the documentation as well.

**Comment by vpnagraj** [comment]
@nistara awesome thanks for working on that ... i totally agree that the default behavior of having all columns displayed as labels is cumbersome 

just a head up ... i see that `graph-annot` is a little behind the master branch and may not include the commit that i issued that for the graph annot argument updates:  21e653c

sorry for any confusion there ... but shouldn't be a problem at all for us to resolve together once you submit a PR

**Comment by nistara** [comment]
@vpnagraj, there are now two branches, the first, 'graph-annot', which prints all columns by default, and the second, 'specific-annot', which prints only the "id" and "group" columns unless other specified. I'm leaving both options so you can see the default behaviour. I personally recommend the 'specific-annot' option!

**Comment by nistara** [comment]
And I also updated the authors for this function to include @thibautjombart and @vpnagraj , because I borrowed heavily from the other `plot` function to make the two comparable.

**Comment by vpnagraj** [comment]
@nistara i took a look your changes and resolved the conflicts with the master branch: dbb4d19  

but it looks like there's an error in `devtools::check()`

```
* creating vignettes ... ERROR
Warning in subset.epicontacts(merskor15, node.attribute = list(outcome = "Dead",  :
  No subsetting attributes provided, returning input object
Quitting from lines 166-167 (epicontacts.Rmd) 
Error: processing vignette 'epicontacts.Rmd' failed with diagnostics:
undefined columns selected
Execution halted
Error: Command failed (1)
Execution halted

Exited with status 1.
```

**Comment by nistara** [comment]
@vpnagraj, build() and check() are passing on my end (on the `specific-annot` branch, before any merge). I'll try and figure out what's causing the issue...

**Comment by nistara** [comment]
@vpnagraj, it appears that the merge nixed my recent changes to the function(!), so it's definitely going to throw errors.

**Comment by vpnagraj** [comment]
aha! @nistara thanks for fixing and merging

**Comment by vpnagraj** [comment]
@nistara and @thibautjombart quick observation ...

i noticed `annot = FALSE` and `annot = NULL` for `vis_epicontacts` / `graph3D` both return error messages:

```
library(outbreaks)
library(epicontacts)
x <- make_epicontacts(linelist = mers_korea_2015$linelist,
                              contacts = mers_korea_2015$contacts, 
                              directed = FALSE)

graph3D(x, annot = FALSE)
graph3D(x, annot = NULL)

vis_epicontacts(x, annot = FALSE)
vis_epicontacts(x, annot = NULL)
```

i guess i was expecting either of those option to work but return a plot without the node labels ... can't decide if that's useful ... what do you guys think?

**Comment by nistara** [comment]
@vpnagraj , I think that's very useful actually. I'll modify the function to work with these `annot` options.

**Comment by nistara** [comment]
Will be able to do it tomorrow night/Friday due to another concurrent deadline. Sorry for the delay...

**Comment by thibautjombart** [comment]
Thanks for flagging this. Should be fixed in `vis_epicontacts` via 77e674108e9bad1737a693c3a5b13371816106a9

**Comment by thibautjombart** [comment]
@nistara catching up with comments: totally agree if `graph3D` becomes bad with all annotations, it's okay to have different behaviours between the functions. Could even enforce a single node attribute to be displayed if you think it useful. up to you :)

**Comment by nistara** [comment]
@thibautjombart, thanks! I've made the default display `id` and grouping variable (if any). If the user wants more, he/she can add as they prefer. I've updated the `graph3D()` function to align with the changes you put in with `vis_epicontacts()`. This has been great for me actually, because I got to learn more nuances of coding from you!!

**Comment by nistara** [comment]
I need to learn about code coverage now. We're down to 98%, and I don't quite know yet how/why.

**Comment by vpnagraj** [comment]
@nistara thanks for all the contributions ... i've made one minor change to accommodate for `group = NULL` or `group = FALSE` having default node colors 396997b

as for the code coverage % ... that's calculated as the n lines of code covered in tests / n lines of total code ... the codecov.io integration gives some helpful visuals / diagnostics for what new code has been introduced and what tests need to be written: <https://codecov.io/github/reconhub/epicontacts?branch=master>

so it's best to make sure that as you're adding new lines of code to the package, those features are covered in the tests/testthat scripts

**Comment by nistara** [comment]
@vpnagraj Thanks for explaining that! I'm going to have to learn about testthat going forward. As of now (after updating `get_clusters()`) I haven't added anything to the tests/testthat script for fear of messing things up.

**Comment by thibautjombart** [comment]
Note that  you can check coverage locally using `covr::shine(covr::package_coverage())`.
:)
